docs/architecture/adr-032-typed-events.md

```diff
@@ -0,0 +1,313 @@
+		return err
+	}
+
+	var attrMap map[string]json.RawMessage
+	err = json.Unmarshal(evtJSON, &attrMap)
+	if err != nil {
+		return err
+	}
+
+	var attrs []abci.EventAttribute
+	for k, v := range attrMap {
+		attrs = append(attrs, abci.EventAttribute{
+			Key:   []byte(k),
+			Value: v,
+		})
+	}
+
+	em.EmitEvent(Event{
+		Type:       evtType,
+		Attributes: attrs,
+	})
+
+	return nil
+}
+
+// ParseTypedEvent converts abci.Event back to typed event
+func ParseTypedEvent(event abci.Event) (proto.Message, error) {
+	concreteGoType := proto.MessageType(event.Type)
+	if concreteGoType == nil {
+		return nil, fmt.Errorf("failed to retrieve the message of type %q", event.Type)
+	}
+
+	var value reflect.Value
+	if concreteGoType.Kind() == reflect.Ptr {
+		value = reflect.New(concreteGoType.Elem())
+	} else {
+		value = reflect.Zero(concreteGoType)
+    }
+    
+	protoMsg, ok := value.Interface().(proto.Message)
+	if !ok {
+		return nil, fmt.Errorf("%q does not implement proto.Message", event.Type)
+	}
+
+	attrMap := make(map[string]json.RawMessage)
+	for _, attr := range event.Attributes {
+		attrMap[string(attr.Key)] = attr.Value
+	}
+
+	attrBytes, err := json.Marshal(attrMap)
+	if err != nil {
+		return nil, err
+	}
+
+	err = jsonpb.Unmarshal(strings.NewReader(string(attrBytes)), protoMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	return protoMsg, nil
+}
+```
+
+Here, the `EmitTypedEvent` is a method on `EventManager` which takes typed event as input and apply json serialization on it. Then it maps the JSON key/value pairs to `event.Attributes` and emits it in form of `sdk.Event`. `Event.Type` will be the type URL of the proto message.
+
+When we subscribe to emitted events on the tendermint websocket, they are emitted in the form of an `abci.Event`. `ParseTypedEvent` parses the event back to it's original proto message. 
+
+__Step-2__: Add proto definitions for typed events for msgs in each module:
+
+For example, let's take `MsgSubmitProposal` of `gov` module and implement this event's type.
+
+```protobuf
+// proto/cosmos/gov/v1beta1/gov.proto
+// Add typed event definition
+
+package cosmos.gov.v1beta1;
+
+message EventSubmitProposal {
+    string from_address   = 1;
+    uint64 proposal_id    = 2;
+    TextProposal proposal = 3;
+}
+```
+
+__Step-3__: Refactor event emission to use the typed event created and emit using `sdk.EmitTypedEvent`:
+
+```go
+// x/gov/handler.go
+func handleMsgSubmitProposal(ctx sdk.Context, keeper keeper.Keeper, msg types.MsgSubmitProposalI) (*sdk.Result, error) {
+    ...
+    types.Context.EventManager().EmitTypedEvent(
+        &EventSubmitProposal{
+            FromAddress: fromAddress,
+            ProposalId: id,
+            Proposal: proposal,
+        },
+    )
+    ...
+}
+```
+
+#### How to subscribe to these typed events in `Client`
+
+> NOTE: Full code example below
+
+Users will be able to subscribe using `client.Context.Client.Subscribe` and consume events which are emitted using `EventHandler`s.
```

Comment:
Can this just be `client.Context.Subscribe` or something?

Reply:
This should be `client.Context.Client.Subscribe`. `Client` in `client.Context` is of type `rpcclient.Client` and  we are using it to `Subscribe`.